stablehlo/transforms/StablehloRefineShapes.cpp

```diff
@@ -0,0 +1,937 @@
+    // structured bindings in a lambda is a C++ 20 extension.
+    auto currentReturnType = std::get<0>(it);
+    auto refinement = std::get<1>(it);
+    auto refinedType = hlo::inferMostSpecificType(
+        /*location=*/{}, {currentReturnType, refinement});
+    if (failed(refinedType))
+      return rewriter.notifyMatchFailure(op, [&](Diagnostic& diag) {
+        diag << "inferMostSpecificType failed for " << currentReturnType
+             << " and " << refinement;
+      });
+    refinedTypes.push_back(*refinedType);
+    needsRefinement |= (currentReturnType != *refinedType);
+  }
+  if (!needsRefinement)
+    return rewriter.notifyMatchFailure(op, "doesn't need refinement");
+
+  for (auto it : llvm::zip(values, refinedTypes)) {
+    // Cannot use structured bindings to simplify this because capturing
+    // structured bindings in a lambda is a C++ 20 extension.
+    auto value = std::get<0>(it);
+    auto refinedType = std::get<1>(it);
+    if (value.getType() == refinedType) continue;
+
+    // Check whether the users of this value are ready for the type of the
+    // value to be refined.
+    for (Operation* user : value.getUsers()) {
+      // CHLO and StableHLO ops are designed to support type refinements of
+      // their operands and results. Any operand type in these ops can change
+      // within what's supported by `inferMostSpecificType` without breaking
+      // verification of the op.
+      if (isa<chlo::ChloDialect, StablehloDialect>(user->getDialect()))
+        continue;
+
+      // Simply changing operand type of `func.return` won't work because
+      // that won't update the FunctionType of the enclosing `func.func`.
+      // Nonetheless, we still want to support these ops because they are widely
+      // used in StableHLO programs (although the plan of record is to replace
+      // `func.return` ops in StableHLO programs with `stablehlo.return`:
+      // https://github.com/openxla/stablehlo/issues/425).
+      if (isa<func::ReturnOp>(user)) continue;
+
+      // Unlike in TensorFlow's type inference pass, here we work only with
+      // allowlisted ops to focus our support on well-defined semantics of
+      // StableHLO programs.
+      return rewriter.notifyMatchFailure(op, [&](Diagnostic& diag) {
+        diag << "unsupported refinement: tried to refine " << value.getType()
+             << " to " << refinedType << " for user " << user;
+      });
+    }
+
+    // Happy path: simply call setType here because most of our users are
+    // fine with that.
+    auto unrefinedType = value.getType();
+    value.setType(refinedType);
+
+    // Special case: for `func.return`, guard the refinement with a
+    // `tensor.cast` and leave propagation of the refined return type to a
+    // dedicated pattern.
+    auto isFuncReturn = [](OpOperand& use) -> bool {
+      return isa<func::ReturnOp>(use.getOwner());
+    };
+    if (llvm::none_of(value.getUses(), isFuncReturn)) continue;
```

Comment:
Question: does this mean that we only do the refinement until a "return", i.e., the end of a function? I think so because the plan in the design doc says that the refinement only works within a mlir function. Just ask to confirm.

Reply:
`applyPatternsAndFoldGreedily` goes through all the ops in the module. For each op, if appropriate it refines its type and propagates it to its uses. The supported uses include CHLO dialect, StableHLO dialect as well as `func.return` because of the fundamental role that it plays. However, because `func.return` is outside of our control, it needs special treatment.

Overall, yes, `func.return` does need special treatment, but not because of our traversal strategy - it's because it is the only op outside of our dialects that we support. Please let me know if this answers the question!

Reply:
Yes, and thanks!